### ISSUE TITLE
How to declare float values in asmjit ?

### ISSUE THREAD

**Issue by thiagomacg** [asmjit/asmjit#261]
Good night,

I m working in a project where i need to represent objective functions in asmjit. Following this example with integer variables. 

        `JitRuntime rt;                          // Runtime specialized for JIT code execution.
        CodeHolder code;                        // Holds code and relocation information.
        code.init(rt.codeInfo());               // Initialize to the same arch as JIT runtime.
        typedef int (*Func)(void);

        x86::Assembler a(&code);                // Create and attach x86::Assembler to `code`.
        a.mov(x86::eax, 1) ;
        a.mov(x86::ebx, 20);
        a.add(x86::eax, x86::ebx);
        a.ret();
`
Already tried some examples available on the site, but was unsuccessful
How can I represent real variables and their operations? 
I need to do something like. 

`       x86::Assembler a(&code);                // Create and attach x86::Assembler to `code`.
        a.mov(x86::eax, 1.5) ;
        a.mov(x86::ebx, 21.5 );
        a.add(x86::eax, x86::ebx);
        a.ret(); `

**Comment by kobalicek** [comment]
Hi, you would have to do the same thing compilers are doing when emitting code, which means either having a data section or embedding constants after / inside your functions. Additionally, you can use asmjit's `ConstPool` or built-in functionality of Builder/Compiler, which wraps `ConstPool` and provides easier API to use it.

RAW `ConstPool` example would look like this:

```c++
#include <asmjit/asmjit.h>

int main() {
  using namespace asmjit;

  JitRuntime rt;
  CodeHolder code;

  code.init(rt.codeInfo());
  x86::Assembler a(&code);

  // ConstPool requires Zone, which must outlive it.
  Zone zone(1024);
  ConstPool constPool(&zone);
  Label constPoolLabel = a.newLabel();

  // Create a constant value of 1.0.
  double one = 1.0;
  size_t oneOffset;
  constPool.add(&one, sizeof(double), oneOffset);

  // Use such constant.
  a.mulps(x86::xmm0, x86::ptr(constPoolLabel, oneOffset));

  // At the end of the code, after it returns, you may embed the data.
  // This would bind the label here with proper alignment and embed
  // all constants.
  a.embedConstPool(constPoolLabel, constPool);

  return 0;
}
```

It's the lowest-level approach, designed to be wrapped into something easier to use.

**Comment by thiagomacg** [comment]
Hi, kobalicek, 
Thanks for you supporting.

I tried to solve it as follows but was unsuccessful. 
The idea is to read symbols of any function iteratively and generate the assembler code.
These symbols can be integer, real or variable (when variable i think i need be stored the variable name to evaluated later).
After finalizing return the function assembler, whose evaluation will be called at another time.

In this code snippet above, I then tried to store the actual values ​​read, perform a sum and return the corresponding assembler function.

What do you suggest ?

        ` JitRuntime tr;
        CodeHolder coder;  
        coder.init(tr.codeInfo());
        x86::Assembler ass(&coder);

        // ConstPool requires Zone, which must outlive it.
        Zone zone(1024);
        ConstPool constPool(&zone);
        Label constPoolLabel = ass.newLabel();

        // Create a constant value of 1.0.
        double one = 1.0;
        size_t oneOffset;
        constPool.add(&one, sizeof(double), oneOffset);

        // Use such constant.
        ass.movups(x86::xmm0, x86::ptr(constPoolLabel, oneOffset));
        // ass.mulps(x86::xmm0, x86::ptr(constPoolLabel, oneOffset));

        constPoolLabel = ass.newLabel();
        double two = 1.5;
        constPool.add(&two, sizeof(double), oneOffset);

        // Use such constant.
        ass.movups(x86::xmm1, x86::ptr(constPoolLabel, oneOffset));
        // ass.mulps(x86::xmm1, x86::ptr(constPoolLabel, oneOffset));

        ass.addps(x86::xmm0, x86::xmm1);

        // At the end of the code, after it returns, you may embed the data.
        // This would bind the label here with proper alignment and embed
        // all constants.
        ass.embedConstPool(constPoolLabel, constPool);
        ass.ret();

        tr.add(&fn, &coder);         // Add the generated code to the runtime.
        // if (err) return 1;                      // Handle a possible error returned by AsmJit.
        // ----> CodeHolder is no longer needed from here and can be destroyed <----

        double result = fn();
        printf("%f\n", result);                 // Print the resulting "2.5". `

**Comment by kobalicek** [comment]
I found several problems in your snippet:
  - Use `movsd` to load a scalar double-precision value, `movups` is not what you want in this case.
  - Use `addsd` if you want to perform addition of two double precision values, movps adds 4 floats.
  - Embed the constant pool AFTER `ret`, you are making all constants part of the code, which would be interpreted as machine code, which is what you don't want.

I have updated my earlier comment to make it more clear
